### DISCUSSION TITLE
How to reduce the memory usage?

### DISCUSSION THREAD

**Discussion by xiaobingxia-at** [milvus-io/milvus#37986] [Q&A and General discussion]
Hello Milvus experts, I'm trying to reduce the memory usage while maintaining a relatively high recall rate, I have tried following options:

- Use in-memory index: IVF-SQ8 (IVF-PQ recall rate is a little bit low), the memory usage is about 40% of raw data size.
- Use HNSW + MMAP: The memory usage is 28% of raw data size. (When hosting 160m 1536-d vectors, raw data is 919GB, query node memory usage is 260GB, and Disk Usage is 610GB)
- Use Disk ANN: with default config, the memory usage is 42.6% of raw data size. (When hosting 134.948m 1536-d vectors, raw data is 772.18GB, disk ann index on disk is: 749.5GB, Query Node memory usage is: 319GB). This is concerning, the reason I come to diskann is because of the limited memory resource. If anyway it will take memory which is >40% of raw data, then there is no difference from using in-memory index. Is there anything I did wrong here?

My goal is to reduce the memory usage to be of 10% of raw data size given my data volume. Is there any good advice?
Thanks!

**Comment by yhmo** [comment]
Typically, the index size compares to the original data size:
- IVF_FLAT/HNSW, a bit larger than the original data size.
- IVF_SQ8/IVF_PQ, 25% ~ 30% of the original data size.
- DISKANN, 1/4 ~ 1/6 of the original data size.

You got 40% memory usage with IVF_SQ8, which is higher than expected. Is your milvus standalone? If so, maybe some compaction/index tasks are not finished.
Normally, if MMAP is enabled, the memory usage is much less than no-mmap. You got 28% for HNSW+MMAP, I think this is expected.
42.6% for DISKANN is also unexpected. Ideally, it should be 20% ~ 25%. Did you change the configuration of PQCodeBugetGBRatio?

**Reply by xiaobingxia-at** [reply]
PQCodeBugetGBRatio is default value 0.125.

**Reply by xiaobingxia-at** [reply]
The memory usage is still relatively large given default PGCodeBugetGBRatio value.

**Comment by xiaofan-luan** [comment]
> Hello Milvus experts, I'm trying to reduce the memory usage while maintaining a relatively high recall rate, I have tried following options:
> 
> * Use in-memory index: IVF-SQ8 (IVF-PQ recall rate is a little bit low), the memory usage is about 40% of raw data size.
> * Use HNSW + MMAP: The memory usage is 28% of raw data size. (When hosting 160m 1536-d vectors, raw data is 919GB, query node memory usage is 260GB, and Disk Usage is 610GB)
> * Use Disk ANN: with default config, the memory usage is 42.6% of raw data size. (When hosting 134.948m 1536-d vectors, raw data is 772.18GB, disk ann index on disk is: 749.5GB, Query Node memory usage is: 319GB). This is concerning, the reason I come to diskann is because of the limited memory resource. If anyway it will take memory which is >40% of raw data, then there is no difference from using in-memory index. Is there anything I did wrong here?
> 
> My goal is to reduce the memory usage to be of 10% of raw data size given my data volume. Is there any good advice? Thanks!

I don't recommend to use IVFSQ/PQ in your use case.

You can tune diskANN params to reduce the memory foot print(disable the cache and increase the quantazitio ratio)
For HNSW I believe you can insert more data until it reaches OOM.

**Reply by xiaobingxia-at** [reply]
How about the IVF-SQ8 + MMAP? I noticed that the search latency is 50ms compared to 10ms when using HNSW + MMAP. But 50ms is good enough for us.

**Reply by xiaobingxia-at** [reply]
And why does increasing quantatizition ratio can help reduce the memory footprint? I thought it will increase the recall by using more memory?

**Reply by xiaofan-luan** [reply]
quantazation change 32bit fp to 8bit, 4bit, 2bit or even 1bit.
the less the bit number, the less memory you will need and potentially better performance , but on the other side lower the recall.

**Reply by xiaofan-luan** [reply]
IVF SQ8 is usually much slower compared to Graph index due to he IO granularity

**Reply by xiaobingxia-at** [reply]
@xiaofan-luan regarding IO granularity..does that mean IVF-SQ8 access the disk more frequently than HNSW?

**Reply by xiaofan-luan** [reply]
> @xiaofan-luan regarding IO granularity..does that mean IVF-SQ8 access the disk more frequently than HNSW?

ideally yes. IVF read the who bucket out of disk(It's sequential IO) but HSNW only read edges or one data out of disk each time(it's random IO). However, compare IVFSQ8 with HNSW will be unfair since the recall will be very different. (IVFSQ8 usually less than 90% while HNSW can be 98-99%)

**Reply by xiaobingxia-at** [reply]
> IVFSQ8 usually less than 90% while HNSW can be 98-99%

This is interesting. I did the recall validation, by using nlist=1024, nprobe=1024 for IVFSQ8 (300k vectors, search top 10, 100, 1000, and compare with the result generated by FLAT index(100% recall)). I got the IVFSQ8's recall rate as 100% or 99.9%.

Do you think that's consistent recall rate for IVF-SQ8 as long as I use nlist=1024 and nprobe=1024?

**Reply by xiaofan-luan** [reply]
> > IVFSQ8 usually less than 90% while HNSW can be 98-99%
> 
> This is interesting. I did the recall validation, by using nlist=1024, nprobe=1024 for IVFSQ8 (300k vectors, search top 10, 100, 1000, and compare with the result generated by FLAT index(100% recall)). I got the IVFSQ8's recall rate as 100% or 99.9%.
> 
> Do you think that's consistent recall rate for IVF-SQ8 as long as I use nlist=1024 and nprobe=1024?

Usuaully nprobe should be much smaller than nlist(if nlist is 1024, nprobe should be set to 32 or 64). With nprobe = 1024, this is more like a brute force search. 
I don't think under nlist=nprobe = 1024 IVFSQ8 will be faster than hnsw, it should be 100 times slower under most cases

**Reply by xiaobingxia-at** [reply]
Thanks @xiaofan-luan ! It is good to know. 
Based on the above information, HNSW is doing a lot of random I/O (one data point each time). Does that mean the disk IOPS will have a lot of impact to HNSW performance? I just found that my /proc/sys/fs/aio-max-nr is 65535, instead of 1m. I assume once I change the aio-max-nr to 1m, my HNSW performance will be immediately improved?

**Reply by xiaobingxia-at** [reply]
Right now, the HNSW search latency is around 300ms - 1s, after stopping the ingestion, and waiting for a few hours, and doing a lot of searches, the latency may reduce to <10ms.
I'm suspecting it is because my ssd disk can only do 65535 aync i/o at the same time.

**Reply by liliu-z** [reply]
proc/sys/fs/aio-max-nr control the concurrency of AIO which is used by DiskANN. But mmap uses another mechanism to do disk interaction, so this parameter cannot affect the performance of mmap.
For better troubleshooting, checking the IOPS number can help. And plz double check the IOPS capability of the path Milvus is running.

**Reply by xiaobingxia-at** [reply]
I did the check for "system.io.w_s" and "system.io.r_s" metrics from datadog. Looks like the write request per second is <1000, and read request per second is very very low like <10.

**Reply by xiaobingxia-at** [reply]
oh, wait, the system.io.r_s was 2800+ if just started the searching process. After ~30 minutes, it gradually went down to 0.
